### PR TITLE
Fix #5059: Backport for 2.3 FACELETS_REFRESH_PERIOD > 0

### LINKED ISSUE #5059: FACELETS_REFRESH_PERIOD > 0 triggers incorrect rendering
**Describe the bug**

If the option `javax.faces.FACELETS_REFRESH_PERIOD` is configured in the `web.xml` to something greater than 0 it makes that some attributes are not rendered OK for the components. In the attached application the HtmlForm loses the `style` attribute.

**To Reproduce**

Compile and deploy the attached app:

```bash
unzip test-jsf.zip
cd test-jsf
mvn clean install
```
Go to the test page, http://127.0.0.1:8080/test-jsf/test.xhtml and click the button two times. The style for the text is different. Check the form in the html page and it has no style attribute.

**Expected behavior**

The form should maintain the style attribute.

**Desktop (please complete the following information):**

 - OS: All
 - Browser All
 - Version All

**Additional context**

Looking the code the problem is that the `DefaultFacelet.java` adds a timestamp when using the refresh time to remove old elements ([here](https://github.com/eclipse-ee4j/mojarra/blob/4.0.0-M5-RELEASE/impl/src/main/java/com/sun/faces/facelets/impl/DefaultFacelet.java#L188-L202)). It happens that the form uses an optimized rendering that renders the attrs in [attributesThatAreSet](https://github.com/eclipse-ee4j/mojarra/blob/master/impl/src/main/java/jakarta/faces/component/UIComponentBase.java#L1893). Those seem to be removed when adding the timestamp.

@arjantijms @BalusC Don't we need to initiate the list of the delta like this?

```diff
diff --git a/impl/src/main/java/jakarta/faces/component/ComponentStateHelper.java b/impl/src/main/java/jakarta/faces/component/ComponentStateHelper.java
index 96d4cd88..e7de3068 100644
--- a/impl/src/main/java/jakarta/faces/component/ComponentStateHelper.java
+++ b/impl/src/main/java/jakarta/faces/component/ComponentStateHelper.java
@@ -210,14 +210,14 @@ class ComponentStateHelper implements StateHelper, TransientStateHelper {
     }
 
     private void initList(Serializable key) {
-        if (component.initialStateMarked()) {
-            deltaMap.computeIfAbsent(key, e -> new ArrayList<>(4));
-        }
-
         if (get(key) == null) {
             List<Object> items = new ArrayList<>(4);
             defaultMap.put(key, items);
         }
+
+        if (component.initialStateMarked()) {
+            deltaMap.computeIfAbsent(key, e -> new ArrayList<>((List<Object>) get(key)));
+        }
     }
```

With that change I'm initializing the delta with the contents of the default, and then adding the new value. If not the delta only contains the added attribute but loses the previous ones. And the restore only resets the timestamp losing the original style.

I can send a PR if you want or you can add the change. I tested with version 2.3 and I see no regression.

Thanks in advance!

[test-jsf.zip](https://github.com/eclipse-ee4j/mojarra/files/8169464/test-jsf.zip)

### PR DESCRIPTION
**Pull request by melloware**
Signed-off-by: melloware <mellowaredev@gmail.com>

### CODE REVIEW

File: impl/src/test/java/javax/faces/component/UIComponentBaseTestCase.java [hunk 1]
```diff
@@ -473,6 +473,21 @@ public void testValueBindings() {
...
+        c.markInitialState();
+        c.getAttributes().put("attr2", "value2");
+        assertEquals(Arrays.asList("attr1", "attr2"), c.getAttributes().get("jakarta.faces.component.UIComponentBase.attributesThatAreSet"));
+
+        Object state = c.saveState(facesContext);
+        c = new ComponentTestImpl();
+        c.pushComponentToEL(facesContext, c);
+        c.restoreState(facesContext, state);
+        c.popComponentFromEL(facesContext);
+        assertEquals(Arrays.asList("attr1", "attr2"), c.getAttributes().get("jakarta.faces.component.UIComponentBase.attributesThatAreSet"));
```
Comment (BalusC) [inline impl/src/test/java/javax/faces/component/UIComponentBaseTestCase.java:488]: Is this valid for 2.3 at all? (javax.faces)

File: impl/src/test/java/javax/faces/component/UIComponentBaseTestCase.java [hunk 2]
```diff
@@ -473,6 +473,21 @@ public void testValueBindings() {
 
     }
 
+    public void testAttributesThatAreSetStateHolder() throws Exception {
+        ComponentTestImpl c = new ComponentTestImpl();
+        c.getAttributes().put("attr1", "value1");
+        c.markInitialState();
+        c.getAttributes().put("attr2", "value2");
+        assertEquals(Arrays.asList("attr1", "attr2"), c.getAttributes().get("jakarta.faces.component.UIComponentBase.attributesThatAreSet"));
```
Comment (BalusC) [inline impl/src/test/java/javax/faces/component/UIComponentBaseTestCase.java:481]: Is this valid for 2.3 at all? (javax.faces)

File: impl/src/test/java/javax/faces/component/UIComponentBaseTestCase.java (hunk 1 above)
Comment (melloware) [reply impl/src/test/java/javax/faces/component/UIComponentBaseTestCase.java]: Sorry updated.  Cherry pickde the commit from the higher branch whoops!

**Comment by BalusC** [comment]
FYI: all 2.3 tests passed

**Comment by melloware** [comment]
nice!

**Pull request merged**